src/ray/raylet/node_manager.cc

```diff
@@ -1672,6 +1673,29 @@ void NodeManager::HandleReturnWorker(const rpc::ReturnWorkerRequest &request,
   send_reply_callback(status, nullptr, nullptr);
 }
 
+void NodeManager::HandleShutdownRaylet(const rpc::ShutdownRayletRequest &request,
+                                       rpc::ShutdownRayletReply *reply,
+                                       rpc::SendReplyCallback send_reply_callback) {
+  RAY_LOG(INFO)
+      << "Shutdown RPC has received. Shutdown will happen after the RPC is replied.";
+  auto graceful = request.graceful();
+  auto shutdown_after_reply = [graceful]() {
```

Comment:
50/50: if it's !graceful, should we just call _Exit?

Reply:
Hmm my impression of _Exit is that it is a hack to avoid bad destruction order. I think raylet has a lot better behavior here (so it is probably not necessary?) up to you